ext/zend_test/test.c

```diff
@@ -73,6 +73,22 @@ static zend_class_entry *zend_test_int_enum;
 static zend_class_entry *zend_test_magic_call;
 static zend_object_handlers zend_test_class_handlers;
 
+static ZEND_FUNCTION(zend_test_gh12986)
+{
+	ZEND_PARSE_PARAMETERS_NONE();
+	HashTable *ht = _zend_new_array_0();
+	HT_ALLOW_COW_VIOLATION(ht);
```

Comment:
If you really need this, can you not initialize the HT before calling `HT_ALLOW_COW_VIOLATION`? Hash tables are very hot, so changes need to be made with caution.

Reply:
@iluuu1994 Yes, that's one of the possible workarounds.

But if the bug can be fixed and benchmarks show no regressions, then why not fix it? 😉

Reply:
> If you really need this, can you not initialize the HT before calling `HT_ALLOW_COW_VIOLATION`?

Ah, unfortunately, `zend_hash_real_init_ex` is not a ZEND_API function...

Reply:
This change may add several instructions + some register pressure (keeping prev_flags alive) for little benefit. This function may be called many thousand times during the handling of a request. If your function relies on `HT_ALLOW_COW_VIOLATION`, it seems better to make it responsible to set the flag in a way that is compatible with the rest of the code.

Reply:
Indeed the benchmarks show a small but noticeable impact.

> If your function relies on HT_ALLOW_COW_VIOLATION, it seems better to make it responsible to set the flag in a way that is compatible with the rest of the code.

Yeah, that's what I'm already doing for my use case.

I was hoping to come-up with a solution so that the next person needing this functionality wouldn't waste time on finding the cause of the issue and reinventing a workaround.

Alas, unless I've missed something, there doesn't seem to be a zero-cost fix for this... 😞

Reply:
> Indeed the benchmarks show a small but noticeable impact.

The benchmark isn't always reliable, but I would consider the diff quite large.

> I was hoping to come-up with a solution so that the next person needing this functionality wouldn't waste time on finding the cause of the issue and reinventing a workaround.

IMO, a comment works well for that. :)

Bob doesn't agree. Dmitry is the primary maintainer of hash maps, so it makes sense to wait for his comment.